File: src/test/kotlin/de/aufmass/kern/PruefungTest.kt
package de.aufmass.kern

import java.math.BigDecimal
import java.util.UUID
import kotlin.math.abs
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/**
 * Die Normprüfung ist das Alleinstellungsmerkmal (Konzept §8). Diese Tests
 * beschreiben die Fälle, die sie melden muss – und die Leitplanke, dass
 * Freiräume immer als untere Schranke herauskommen.
 */
class PruefungTest {

    private fun nah(erwartet: Double, ist: Double, toleranz: Double = 0.005) =
        assertTrue(abs(erwartet - ist) < toleranz, "erwartet $erwartet, war $ist")

    private fun schrankHinten(raum: Raum, art: String, xM: String, tiefeM: Double): Pruefung.Platzhalter {
        // Mit dem Rücken an die hintere Wand (y = Raumtiefe), Vorderseite in den
        // Raum (Drehung 180° → n zeigt nach -y).
        val ht = tiefeM / 2
        return Pruefung.Platzhalter(
            raumId = raum.id, komponente = art,
            xM = BigDecimal(xM), yM = BigDecimal(raum.tiefeM.toDouble() - ht),
            drehungGrad = BigDecimal("180")
        )
    }

    @Test
    fun `Grundflaeche eines Kloetzchens hat die Katalogmasse`() {
        val raum = Raum(name = "Technik", breiteM = BigDecimal("4"), tiefeM = BigDecimal("3"))
        val komp = Komponenten.finde("ZAEHLERSCHRANK")!!    // 0,60 × 0,20
        val ph = schrankHinten(raum, "ZAEHLERSCHRANK", "2.0", 0.20)
        val e = Pruefung.grundriss(ph, komp)
        assertEquals(4, e.size)
        nah(0.60, abs(e[1].x - e[0].x))    // Breite entlang der Wand
        nah(0.20, abs(e[2].y - e[1].y))    // Tiefe in den Raum
        nah(3.0, e.maxOf { it.y })          // Rücken an der hinteren Wand
    }

    @Test
    fun `genug Platz vor dem Zaehlerschrank ist in Ordnung`() {
        val raum = Raum(name = "Technik", breiteM = BigDecimal("4"), tiefeM = BigDecimal("3"))
        val ph = schrankHinten(raum, "ZAEHLERSCHRANK", "2.0", 0.20)
        val befunde = Pruefung.pruefe(raum, listOf(ph))
        val bedien = befunde.single { it.regelId == "BEDIENBEREICH_VERTEILUNG" }
        assertEquals(Pruefung.Schwere.OK, bedien.schwere)
        // ~2,80 m frei, konservativ 2,78 – jedenfalls über den geforderten 1,20 m.
        assertTrue(bedien.verfuegbarM!! >= BigDecimal("1.20"))
    }

    @Test
    fun `zu wenig Freiraum vor der Verteilung wird gewarnt`() {
        // Flacher Raum: nur 1,20 m tief – der Bedienbereich passt nicht davor.
        val raum = Raum(name = "Nische", breiteM = BigDecimal("4"), tiefeM = BigDecimal("1.20"))
        val ph = schrankHinten(raum, "ZAEHLERSCHRANK", "2.0", 0.20)
        val bedien = Pruefung.pruefe(raum, listOf(ph)).single { it.regelId == "BEDIENBEREICH_VERTEILUNG" }
        assertEquals(Pruefung.Schwere.WARNUNG, bedien.schwere)
        // 1,00 m roh − 2 cm Toleranz, abgerundet = 0,98 m.
        assertEquals(BigDecimal("0.98"), bedien.verfuegbarM)
        assertTrue(bedien.verfuegbarM!! < bedien.erforderlichM)
    }

    @Test
    fun `Freiraum ist immer die untere Schranke`() {
        // Konservativ: gemessen minus Toleranz, auf ganze cm abgerundet.
        assertEquals(BigDecimal("2.78"), Pruefung.konservativ(2.80))
        assertEquals(BigDecimal("1.18"), Pruefung.konservativ(1.209))
        assertEquals(BigDecimal("0.00"), Pruefung.konservativ(0.01))
    }

    @Test
    fun `Waermequelle zu nah an einem Schrank wird gewarnt`() {
        val raum = Raum(name = "Heizraum", breiteM = BigDecimal("4"), tiefeM = BigDecimal("3"))
        val therme = Pruefung.Platzhalter(
            raumId = raum.id, komponente = "THERME",
            xM = BigDecimal("1.0"), yM = BigDecimal("1.0")
        )
        val schrank = Pruefung.Platzhalter(
            raumId = raum.id, komponente = "SCHALTSCHRANK",
            xM = BigDecimal("1.7"), yM = BigDecimal("1.0")
        )
        val befund = Pruefung.pruefe(raum, listOf(therme, schrank))
            .single { it.regelId == "ABSTAND_WAERMEQUELLE" }
        assertEquals(Pruefung.Schwere.WARNUNG, befund.schwere)
        assertTrue(befund.verfuegbarM!! < BigDecimal("0.30"))
        assertTrue(befund.text.contains("Schaltschrank"))
    }

    @Test
    fun `Kabel bekommt einen Biegeradius-Hinweis`() {
        val raum = Raum(name = "Technik", breiteM = BigDecimal("4"), tiefeM = BigDecimal("3"))
        val kabel = Pruefung.Platzhalter(
            raumId = raum.id, komponente = "KABEL_NYM5",       // Ø 13 mm, 15×
            xM = BigDecimal("2.0"), yM = BigDecimal("1.5")
        )
        val hinweis = Pruefung.pruefe(raum, listOf(kabel)).single { it.regelId == "BIEGERADIUS" }
        assertEquals(Pruefung.Schwere.HINWEIS, hinweis.schwere)
        assertTrue(hinweis.text.contains("195 mm"))
    }

    @Test
    fun `jeder Befund traegt die Fassung der Regelbasis`() {
        val raum = Raum(name = "Technik", breiteM = BigDecimal("4"), tiefeM = BigDecimal("3"))
        val ph = schrankHinten(raum, "ZAEHLERSCHRANK", "2.0", 0.20)
        Pruefung.pruefe(raum, listOf(ph)).forEach {
            assertEquals(Regelwerk.version, it.regelVersion)
        }
    }

    @Test
    fun `unbekannte Komponente wird uebersprungen, nicht geworfen`() {
        val raum = Raum(name = "Technik", breiteM = BigDecimal("4"), tiefeM = BigDecimal("3"))
        val ph = Pruefung.Platzhalter(
            raumId = raum.id, komponente = "GIBTESNICHT",
            xM = BigDecimal("2.0"), yM = BigDecimal("1.5")
        )
        assertEquals(emptyList(), Pruefung.pruefe(raum, listOf(ph)))
    }
}

File: web/tests/pruefung.test.mjs
/*
 * Prüft, dass der JS-Spiegel der Normprüfung dieselben Zahlen liefert wie der
 * Kotlin-Kern (dessen Werte in PruefungTest.kt festgeschrieben sind). Ohne DOM
 * lauffähig – die geprüften Module sind rein.
 *
 * Aufruf: node web/tests/pruefung.test.mjs   (Exitcode != 0 bei Fehler)
 */
import { pruefe, konservativ, grundriss } from "../pruefung.js";
import { finde } from "../komponenten.js";

let ok = 0, fail = 0;
const eq = (name, got, want) => {
  const p = (typeof got === "number" && typeof want === "number") ? Math.abs(got - want) < 1e-9 : got === want;
  if (!p) { console.error(`FAIL ${name}: got=${got} want=${want}`); fail++; } else ok++;
};

// Konservative untere Schranke
eq("konservativ(2.80)", konservativ(2.80), 2.78);
eq("konservativ(1.209)", konservativ(1.209), 1.18);
eq("konservativ(0.01)", konservativ(0.01), 0.0);

// Grundfläche eines Klötzchens = Katalogmaße
const raum = { id: "r", breiteM: 4, tiefeM: 3, hoeheM: 2.5 };
const phZ = { id: "z", komponente: "ZAEHLERSCHRANK", xM: 2.0, yM: 2.9, drehungGrad: 180 };
const e = grundriss(phZ, finde("ZAEHLERSCHRANK"));
eq("grundriss breite", Math.round(Math.abs(e[1][0] - e[0][0]) * 100) / 100, 0.6);
eq("grundriss tiefe", Math.round(Math.abs(e[2][1] - e[1][1]) * 100) / 100, 0.2);

// Genug Platz -> OK
let b = pruefe(raum, [phZ]).find(x => x.regelId === "BEDIENBEREICH_VERTEILUNG");
eq("bedien OK", b.schwere, "OK");
eq("bedien verf 2.78", b.verfuegbarM, 2.78);

// Flacher Raum -> Warnung
const raum2 = { id: "r2", breiteM: 4, tiefeM: 1.2, hoeheM: 2.5 };
const phZ2 = { id: "z2", komponente: "ZAEHLERSCHRANK", xM: 2.0, yM: 1.1, drehungGrad: 180 };
b = pruefe(raum2, [phZ2]).find(x => x.regelId === "BEDIENBEREICH_VERTEILUNG");
eq("flach WARNUNG", b.schwere, "WARNUNG");
eq("flach verf 0.98", b.verfuegbarM, 0.98);

// Wärmequelle zu nah
const therme = { id: "t", komponente: "THERME", xM: 1.0, yM: 1.0, drehungGrad: 0 };
const schrank = { id: "s", komponente: "SCHALTSCHRANK", xM: 1.7, yM: 1.0, drehungGrad: 0 };
b = pruefe(raum, [therme, schrank]).find(x => x.regelId === "ABSTAND_WAERMEQUELLE");
eq("waerme WARNUNG", b.schwere, "WARNUNG");
eq("waerme text Schaltschrank", b.text.includes("Schaltschrank"), true);

// Biegeradius-Hinweis
const kabel = { id: "k", komponente: "KABEL_NYM5", xM: 2, yM: 1.5, drehungGrad: 0 };
b = pruefe(raum, [kabel]).find(x => x.regelId === "BIEGERADIUS");
eq("biegeradius HINWEIS", b.schwere, "HINWEIS");
eq("biegeradius 195mm", b.text.includes("195 mm"), true);

// Unbekannte Komponente -> übersprungen
eq("unbekannt leer", pruefe(raum, [{ id: "x", komponente: "GIBTESNICHT", xM: 2, yM: 1.5 }]).length, 0);

console.log(`pruefung.test: ${ok} ok, ${fail} fail`);
process.exit(fail ? 1 : 0);
